File: platforms/android/lib/src/main/java/com/shopify/checkoutkit/HttpResponseHandler.kt
package com.shopify.checkoutkit

internal class HttpResponseHandler {

    internal enum class Disposition {
        HANDLE_NORMALLY,
        RENDER,
        DISCARD_PRELOAD,
    }

    internal fun disposition(
        responseHeaders: Map<String, String>?,
        isForMainFrame: Boolean,
        isBackgroundedPreload: Boolean,
    ): Disposition {
        if (!isManagedChallenge(responseHeaders)) {
            return Disposition.HANDLE_NORMALLY
        }

        return if (isForMainFrame && isBackgroundedPreload) {
            Disposition.DISCARD_PRELOAD
        } else {
            Disposition.RENDER
        }
    }

    private fun isManagedChallenge(responseHeaders: Map<String, String>?): Boolean {
        val mitigation = responseHeaders
            ?.entries
            ?.firstOrNull { (name) -> name.equals(MITIGATION_HEADER, ignoreCase = true) }
            ?.value
            ?.trim()

        return mitigation.equals(CHALLENGE_VALUE, ignoreCase = true)
    }

    private companion object {
        private const val MITIGATION_HEADER = "cf-mitigated"
        private const val CHALLENGE_VALUE = "challenge"
    }
}

File: platforms/swift/Sources/ShopifyCheckoutKit/HTTPResponseHandler.swift
import Foundation

final class HTTPResponseHandler {
    enum Disposition: Equatable {
        case handleNormally
        case render
        case discardPreload
    }

    func disposition(
        for response: HTTPURLResponse,
        isForMainFrame: Bool,
        isBackgroundedPreload: Bool
    ) -> Disposition {
        guard isManagedChallenge(response) else {
            return .handleNormally
        }

        guard isForMainFrame, isBackgroundedPreload else {
            return .render
        }

        return .discardPreload
    }

    private func isManagedChallenge(_ response: HTTPURLResponse) -> Bool {
        response.value(forHTTPHeaderField: "cf-mitigated")?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .caseInsensitiveCompare("challenge") == .orderedSame
    }
}
